### PR TITLE
[wasm] Wasm TimeZoneInfo implemenation

### PR DESCRIPTION
**Pull request by kjpou1**
Right now WebAssembly does not have an implementation for `TimeZoneInfo` and instead throws exceptions.

__BCL Changes:__

- Add method implementations to TimeZoneInfo.WebAssembly.cs
- Add icall definitions for:
   - mono_timezone_get_data
   - mono_timezone_get_local_name
   - mono_timezone_get_names

__WASM Build__

- Add targets to run time zone tests.
   - make run-tzd-test
   - make run-tzd-interp-sample
   - make run-tzd-aot-sample

__WASM Driver - aka `driver.c`__

- Add implementations of the following:
   - mono_timezone_get_data
   - mono_timezone_get_names
   - mono_timezone_get_local_name

These will be called from the managed bcl module `TimeZoneInfo.WebAssembly.cs` module.

__Zone Info__
javascript module that supports the IANA time zone information is hosted temporarily at https://kjpou1.github.io/zoneinfo/1.0.0/mono-webassembly-zoneinfo.js

`mono-webassembly-zoneinfo.js` module generation repo temporarily at:  https://github.com/kjpou1/mono-webassembly-zoneinfo

Other options to consider:
   - Adding the database implementation to our mono wasm runtime that is built.  This will add ~1 MB to the size though.

### CODE REVIEW

File: sdks/wasm/src/driver.c [hunk 1]
```diff
@@ -674,3 +680,146 @@ mono_wasm_enable_on_demand_gc (void)
 {
 	mono_wasm_enable_gc = 1;
 }
+
```
Comment (vargaz) [inline sdks/wasm/src/driver.c]: This should follow the existing coding conventions in this file.

File: sdks/wasm/src/driver.c [hunk 2]
```diff
@@ -77,6 +81,8 @@ m_strdup (const char *str)
 
 static MonoDomain *root_domain;
 
+static int ZONE_INFO_FIRST_TIME = 1;
+
```
Comment (vargaz) [inline sdks/wasm/src/driver.c]: Its better to have 0 as the default value, so it doesn't need to be initialized. Also, variable names should be lowercase.

File: sdks/wasm/src/driver.c [hunk 3]
```diff
@@ -674,3 +680,146 @@ mono_wasm_enable_on_demand_gc (void)
...
+		//fprintf (stdout, "%s\n", "We have class");
+	    static MonoMethod* method;
+		MonoException* exc = NULL;
+
+    	if (!method)
+    	{
+        	method = mono_class_get_method_from_name(zoneInfoClass, "mono_timezone_get_data", -1);
+    	}
+		
+		void* args[] = {name, size};
```
Comment (vargaz) [inline sdks/wasm/src/driver.c]: The indentation is messed up.

Comment (kjpou1) [reply sdks/wasm/src/driver.c]: I do believe this was already fixed

File: mono/metadata/icall-decl.h [hunk 4]
```diff
@@ -198,6 +198,13 @@ ICALL_EXPORT gint32 ves_icall_System_IO_Compression_DeflateStreamNative_Flush (g
 ICALL_EXPORT gint32 ves_icall_System_IO_Compression_DeflateStreamNative_ReadZStream (gpointer stream, gpointer buffer, gint32 length);
 ICALL_EXPORT gint32 ves_icall_System_IO_Compression_DeflateStreamNative_WriteZStream (gpointer stream, gpointer buffer, gint32 length);
 #endif
+
+#if defined(TARGET_WASM)
+ICALL_EXPORT gpointer ves_icall_System_TimeZoneInfo_xamarin_timezone_get_data (MonoString* name, gint32 *size);
```
Comment (jaykrell) [inline mono/metadata/icall-decl.h]: The first and last should use handles, or perhaps other means.
For example, the first could use `fixed`.
I also suggest `ref to ref` but it isn't yet a widespread idiom.
Either way, the first and last are "wrong".

File: mono/metadata/icall-def.h [hunk 5]
```diff
@@ -1104,6 +1104,13 @@ ICALL_TYPE(WAITH, "System.Threading.WaitHandle", WAITH_1)
 HANDLES(WAITH_1, "SignalAndWait_Internal", ves_icall_System_Threading_WaitHandle_SignalAndWait_Internal, gint32, 3, (gpointer, gpointer, gint32))
 HANDLES(WAITH_2, "Wait_internal", ves_icall_System_Threading_WaitHandle_Wait_internal, gint32, 4, (gpointer_ptr, gint32, MonoBoolean, gint32))
 
+#if defined(TARGET_WASM)
+ICALL_TYPE(TZINFO, "System.TimeZoneInfo", TZINFO_1)
+NOHANDLES(ICALL(TZINFO_1, "xamarin_timezone_get_data", ves_icall_System_TimeZoneInfo_xamarin_timezone_get_data))
+NOHANDLES(ICALL(TZINFO_2, "xamarin_timezone_get_local_name", ves_icall_System_TimeZoneInfo_xamarin_timezone_get_local_name))
+NOHANDLES(ICALL(TZINFO_3, "xamarin_timezone_get_names", ves_icall_System_TimeZoneInfo_xamarin_timezone_get_names))
```
Comment (jaykrell) [inline mono/metadata/icall-def.h]: Relates to previous comment -- `HANDLES()`, etc.

File: mono/metadata/icall.c [hunk 6]
```diff
@@ -8121,6 +8121,44 @@ ves_icall_System_IO_Compression_DeflateStreamNative_WriteZStream (gpointer strea
...
+#if defined(TARGET_WASM) 
+
+G_EXTERN_C gpointer xamarin_timezone_get_data (MonoString* name, gint32 *size);
+G_EXTERN_C gpointer xamarin_timezone_get_names (gint32 *count);
+G_EXTERN_C MonoString* xamarin_timezone_get_local_name (void);
+
+gpointer
+ves_icall_System_TimeZoneInfo_xamarin_timezone_get_data (MonoString* name, gint32 *size)
+{
+#ifdef MONO_CROSS_COMPILE
```
Comment (jaykrell) [inline mono/metadata/icall.c]: Why cross compile does not work? Put in a comment?

File: sdks/wasm/Makefile [hunk 7]
```diff
@@ -324,6 +327,43 @@ $(eval $(call XunitTestTemplate,System.Xml-xunit,System.Xml-xunit,wasm_System.Xm
...
+# Targets for building time zone data
+#
+
+
+$(TZ_INPUT)/.stamp-tz-data:
+	mkdir -p $(TZ_INPUT)
+	mkdir -p $(TZ_OUTPUT)
+	curl -L https://data.iana.org/time-zones/tzdata-latest.tar.gz -o $(TZ_INPUT)/tzdata.tar.gz
+	tar xvzf $(TZ_INPUT)/tzdata.tar.gz -C $(TZ_INPUT)
+	touch $@
```
Comment (jaykrell) [inline sdks/wasm/Makefile]: Is it possible to do this correctly incrementally? I suspect not.
You know, like, only go to the network when needed? Neither too much nor too little?
Can you use curl to get just a checksum first?

File: sdks/wasm/src/driver.c (hunk 2 above)
Comment (jaykrell) [reply sdks/wasm/src/driver.c]: `gboolean`?
`char` for size?

File: sdks/wasm/src/driver.c [hunk 8]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
...
+
+    return zoneInfoAssembly;
+}
+
+MonoClass* GetZoneInfoClass()
+{
+	static MonoClass* klass;
+	if (!klass)
+		klass = mono_class_from_name(mono_assembly_get_image(GetZoneInfoAssembly()), "WebAssembly.ZoneInfo", "MonoWasmZoneInfo");
+	return klass;
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Kinda needs barriers.

File: sdks/wasm/src/driver.c (hunk 3 above)
Comment (jaykrell) [reply sdks/wasm/src/driver.c]: Needs barriers.

File: sdks/wasm/src/driver.c [hunk 9]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
...
+
+		var buff = Module._malloc((res.length + 1) * 2);
+		stringToUTF16 (res, buff, (res.length + 1) * 2);
+		return buff;
+	}
+})
+
+
+MonoString*
+xamarin_timezone_get_local_name ()
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: `(void)`

File: sdks/wasm/src/driver.c [hunk 10]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
...
+{
+	mono_unichar2 *tzd_local_name = (mono_unichar2*)mono_wasm_timezone_get_local_name();
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Space before open paren in function call. (I won't point that out every time.)

File: sdks/wasm/src/driver.c [hunk 11]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
...
+
+	MonoClass *zoneInfoClass = GetZoneInfoClass(); 
+	if (zoneInfoClass)
+	{
+		static MonoMethod* method;
+		MonoException* exc = NULL;
+
+		if (!method)
+		{
+			method = mono_class_get_method_from_name(zoneInfoClass, "mono_timezone_get_names", -1);
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Needs memory barriers.

File: sdks/wasm/src/driver.c [hunk 12]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
...
+		}
+		
+		void* args[] = {count};
+		MonoObject* ret = mono_runtime_invoke(method, NULL, args, (MonoObject**)&exc);
+		if (!*count)
+			return NULL;
+
+		if(!exc)
+		{
+			MonoArray* names =(MonoArray*)ret;
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Some coop handles in new code?

File: sdks/wasm/src/driver.c [hunk 13]
```diff
@@ -47,6 +47,10 @@ static MonoClass* datetimeoffset_class;
 
 int mono_wasm_enable_gc;
 
+MONO_API void*  xamarin_timezone_get_data (MonoString* name, int *size);
+MONO_API char**  xamarin_timezone_get_names (int *count);
+MONO_API MonoString*  xamarin_timezone_get_local_name (void);
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Is `xamarin_` a new prefix, vs. `mono_`?
Put in a header?

File: sdks/wasm/src/driver.c (hunk 2 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: Yeah, not sure why I did 1 and not zero.

File: sdks/wasm/src/driver.c (hunk 1 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: Fixed in commit

File: mono/metadata/icall.c (hunk 6 above)
Comment (kjpou1) [reply mono/metadata/icall.c]: Addressed in commit

File: sdks/wasm/src/driver.c [hunk 14]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
 {
 	mono_wasm_enable_gc = 1;
 }
+
+MonoAssembly* get_zone_Info_assembly()
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Symbols should start `mono_` or be file-level `static`.
Some other prefixes are ok too, like `ves_icall_` or maybe this new `xamarin_` thing.
Functions that take no parameters should be `(void)`, not `()`, until/unless it is C++.
`()` in C means "accepts anything", `(void)` means "accepts nothing".
They mean the same thing in C++, and `(void)` means the same thing in both languages.

File: sdks/wasm/src/driver.c [hunk 15]
```diff
@@ -674,3 +680,142 @@ mono_wasm_enable_on_demand_gc (void)
...
+	if (!try_load_zone_info) {
+
+		fprintf (stdout, "%s\n", "Locating ZoneInfo assembly.");
+		// try to open the ZoneInfo assembly
+		zoneInfoAssembly = mono_domain_assembly_open (root_domain, "WebAssembly.ZoneInfo");
+		// if not open then try to load it.
+		if (!zoneInfoAssembly) {
+			MonoImageOpenStatus status;
+			MonoAssemblyName* aname = mono_assembly_name_new ("WebAssembly.ZoneInfo");
+			zoneInfoAssembly = mono_assembly_load (aname, NULL, &status);
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: On-demand initialization requires memory barriers, both in the read and write paths, to account for multiple threads.

File: sdks/wasm/src/driver.c (hunk 9 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: fixed in recent commit

File: sdks/wasm/src/driver.c (hunk 10 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: should all be fixed

File: sdks/wasm/src/driver.c (hunk 14 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: fixed in recent commit

**Comment by kjpou1** [comment]
@monojenkins build failed

File: mono/metadata/icall-decl.h (hunk 4 above)
Comment (kjpou1) [reply mono/metadata/icall-decl.h]: Using handles requires access to the internal API's.  As tradeoff, think I am going to change this to use `mono_add_internal_call`.  This should wrap everything up nicely.

Comment (jaykrell) [reply mono/metadata/icall-decl.h]: I not sure how "nicely". This is rarely how things are done. But might be ok.
I suggest you avoid handles and their need using other means to solve the same problem:

1. For passing strings, use `fixed` and then `char*` and `int length`. There are a number of examples of this. For example the performance counters. It is quite easy and efficient and a blessed pattern.

2. Or similarly, you can convert to utf8 on the managed size. Pass a byte* then. I don't like it, but it occurs too. Probably requires extra managed/native roundtrips also.

3. Maybe something with p/invoke. I have no experience here but people use it.

3. For anything else non-MonoString, including the return value, pass a ref to local object. That is equivalent to a coop handle. See my PR removing handles from Monitor for example.
https://github.com/mono/mono/pull/17032
It was deemed ok but not actually approved. I don't know why.
Now it has conflicts.

And then you can say you are nohandles() and be coop-friendly.

File: sdks/wasm/Makefile [hunk 16]
```diff
@@ -25,6 +25,9 @@ CSC?= MONO_PATH=$(TOP)/mcs/class/lib/build $(TOP)/sdks/builds/bcl/runtime/mono-w
 CSC_FLAGS := /debug:portable /noconfig /nostdlib /nologo /langversion:latest
 API_REFS=$(TOP)/external/binary-reference-assemblies/v4.6
 MONO_SUPPORT=$(WASM_RUNTIME_DIR)/include/support
+TZ_INPUT=$(TOP)/sdks/wasm/zoneinfodata
+TZ_OUTPUT=$(TOP)/sdks/wasm/zoneinfo
+
```
Comment (vargaz) [inline sdks/wasm/Makefile]: Extra empty line.

File: sdks/wasm/Makefile (hunk 7 above)
Comment (vargaz) [reply sdks/wasm/Makefile]: Its 400k.

File: sdks/wasm/src/driver.c [hunk 17]
```diff
@@ -674,3 +680,145 @@ mono_wasm_enable_on_demand_gc (void)
...
 }
+
+static MonoAssembly* 
+mono_get_zone_Info_assembly (void)
+{
+	static MonoAssembly* zoneInfoAssembly;
+
+	if (!try_load_zone_info) 
+	{
+		fprintf (stdout, "%s\n", "Locating ZoneInfo assembly.");
```
Comment (vargaz) [inline sdks/wasm/src/driver.c]: This should use K&R indentation.

File: sdks/wasm/src/library_mono.js [hunk 18]
```diff
@@ -100,7 +100,7 @@ var MonoSupportLib = {
 		mono_wasm_runtime_ready: function () {
 			console.log ("MONO-WASM: Runtime is ready.");
 			this.mono_wasm_runtime_is_ready = true;
-			debugger;
+			//debugger;
 		},
```
Comment (vargaz) [inline sdks/wasm/src/library_mono.js]: Is this intentional ?

Comment (kjpou1) [reply sdks/wasm/src/library_mono.js]: fixed

**Comment by kjpou1** [comment]
Where is the leak?

**Comment by jaykrell** [comment]
mono_string_to_utf8 result is leaked.
You don’t likely need the strdup.
In either case, you *can* return MonoString.
Try the “star star”? It should fly?

**Comment by jaykrell** [comment]
Or move to metadata. 
Icalls really should not be written this way imho.

**Comment by kjpou1** [comment]
> mono_string_to_utf8 result is leaked.

You are right.  Thought I had changed that.

File: sdks/wasm/Makefile (hunk 16 above)
Comment (kjpou1) [reply sdks/wasm/Makefile]: Removed

**Comment by kjpou1** [comment]
@monojenkins build failed

File: sdks/wasm/Makefile (hunk 7 above)
Comment (kjpou1) [reply sdks/wasm/Makefile]: Has been removed in favor of `mono_webassembly_zoneinfo.js` module

File: sdks/wasm/src/driver.c (hunk 8 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: This is no longer relevant with current implementation.

File: sdks/wasm/src/driver.c (hunk 11 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: This is no longer relevant with current implementation.

File: sdks/wasm/src/driver.c (hunk 12 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: This is no longer relevant with current implementation.

File: sdks/wasm/src/driver.c (hunk 13 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: These are all `mono_` now

Comment (vargaz) [reply sdks/wasm/src/driver.c]: Its ok to name them xamarin, mono is just the prefix for the runtime. Its ok to name them mono as well.

File: sdks/wasm/src/driver.c (hunk 3 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: This is no longer relevant with current implementation.

File: sdks/wasm/src/driver.c (hunk 2 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: This is no longer relevant with current implementation as the flag was removed

**Comment by kjpou1** [comment]
@monojenkins build failed

File: sdks/wasm/src/driver.c [hunk 19]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+		return Mono_WebAssembly_ZoneInfo.mono_timezone_get_data(str, size);
+	}
+	return null;
+})
+
+void*
+mono_timezone_get_data (mono_unichar2 *name, int name_length, int *size)
+{
+
+	MonoString *zi_name;
```
Comment (vargaz) [inline sdks/wasm/src/driver.c]: Extra line.

File: sdks/wasm/tzdtest.cs [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+                    _timeZoneInstance = TimeZoneInfo.FindSystemTimeZoneById(TimeZone);
+                }
+            }
+            return _timeZoneInstance;
+        }
+        private set { _timeZoneInstance = value; }
+    }
+
+    private static TimeZoneInfo _timeZoneInstance;
+
```
Comment (vargaz) [inline sdks/wasm/tzdtest.cs]: Extra lines.

File: sdks/wasm/tzdtest.cs [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        get { return _timeZone; }
+        set
+        {
+            TimeZoneInstance = null;
+            _timeZone = value;
+        } 
+    }
+    private static string _timeZone;
+
```
Comment (vargaz) [inline sdks/wasm/tzdtest.cs]: Fields are private by default.

File: sdks/wasm/tzdtest.cs (hunk 20 above)
Comment (vargaz) [inline sdks/wasm/tzdtest.cs]: Ditto.

File: sdks/wasm/src/driver.c [hunk 22]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+})
+
+char**
+mono_timezone_get_names (int *count)
+{
+	*count = mono_wasm_zoneinfo_timezone_get_name_count ();
+	if (!*count)
+		return NULL;
+
+	char** result = (char**) malloc (sizeof (char*) * (*count));
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: `calloc`

File: sdks/wasm/src/driver.c [hunk 23]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+	int buffer = mono_wasm_zoneinfo_timezone_get_data((int)native_val, size);
+	return (void*)buffer;
+}
+
+// Returns the number of zone id's available.
+EM_JS(int, mono_wasm_zoneinfo_timezone_get_name_count, (), {
+
+	if (typeof Mono_WebAssembly_ZoneInfo !== "undefined")
+	{
+		return Mono_WebAssembly_ZoneInfo.mono_wasm_timezone_get_names_count ();
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Brace on previous line.

File: sdks/wasm/src/driver.c [hunk 24]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+// Returns the number of zone id name at a given index.
+EM_JS(char*, mono_wasm_zoneinfo_timezone_get_name_at_index, (int index), {
+
+	var res = "";
+	if (typeof Mono_WebAssembly_ZoneInfo !== "undefined")
+	{
+		try {
+			return Mono_WebAssembly_ZoneInfo.mono_wasm_timezone_get_name_at_index (index);
+		}
+		catch (wte) {
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Catch on previous line.

File: sdks/wasm/src/driver.c [hunk 25]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+	else
+		zi_name = mono_string_new_utf16(root_domain, name, name_length);
+	// char *native_name = mono_string_to_utf8(zi_name);
+	// fprintf (stdout, "Called mono_timezone_get_data for %s - %i\n", native_name, name_length);
+	// mono_free (native_name);
+	*size = 0;
+	
+	char *native_val = mono_string_to_utf8 (zi_name);
+	int buffer = mono_wasm_zoneinfo_timezone_get_data((int)native_val, size);
+	return (void*)buffer;
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Do not cast pointers and possibly differently-sized integers. Use `gsize` to go between.
Better yet, use `gsize` here in place of `int`, for when wasm becomes 64bit.

Comment (jaykrell) [reply sdks/wasm/src/driver.c]: Feels like you leak `mono_string_to_utf8` but maybe not.

File: sdks/wasm/src/driver.c [hunk 26]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+	MonoString *name = mono_string_from_utf16 (tzd_local_name);
+	free (tzd_local_name);
+	return name;
+}
+
+// Returns the timezone information for a specfic id.
+EM_JS(int, mono_wasm_zoneinfo_timezone_get_data, (int id, int *size), {
+
+	if (typeof Mono_WebAssembly_ZoneInfo !== "undefined")
+	{
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Brace on previous line.

File: sdks/wasm/src/driver.c [hunk 27]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+	// result in cast warnings from the compiler.
+	mono_unichar2 *tzd_local_name = (mono_unichar2*)mono_wasm_timezone_get_local_name ();
+	MonoString *name = mono_string_from_utf16 (tzd_local_name);
+	free (tzd_local_name);
+	return name;
+}
+
+// Returns the timezone information for a specfic id.
+EM_JS(int, mono_wasm_zoneinfo_timezone_get_data, (int id, int *size), {
+
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Possibly brace on next line (since start of function).

File: sdks/wasm/src/driver.c [hunk 28]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
...
+// The following timezone information routines depends on the Mono_WebAssembly_ZoneInfo module being available.
+//
+
+// Returns the local timezone default is UTC.
+EM_JS(int, mono_wasm_timezone_get_local_name, (), {
+	var res = "UTC";
+	try {
+		return Mono_WebAssembly_ZoneInfo.mono_wasm_timezone_get_local_name ()
+	}
+	catch (wte) {
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: Presumably `catch` on previous line, like `else`.

File: sdks/wasm/src/driver.c [hunk 29]
```diff
@@ -674,3 +674,123 @@ mono_wasm_enable_on_demand_gc (void)
 {
 	mono_wasm_enable_gc = 1;
 }
+
+//
+// The following timezone information routines depends on the Mono_WebAssembly_ZoneInfo module being available.
+//
+
+// Returns the local timezone default is UTC.
+EM_JS(int, mono_wasm_timezone_get_local_name, (), {
```
Comment (jaykrell) [inline sdks/wasm/src/driver.c]: `int` or `gsize`? I mean, when wasm is 64bits, and 32bits, what will this be?

File: sdks/wasm/src/driver.c (hunk 25 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: Yes it is.  Thanks for the catch.

File: sdks/wasm/src/driver.c (hunk 29 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: gsize is not defined.

File: sdks/wasm/src/driver.c (hunk 22 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: why `calloc` here?  The full memory space will be used so no need to clear it first unless I am missing something.

File: sdks/wasm/src/driver.c (hunk 25 above)
Comment (kjpou1) [reply sdks/wasm/src/driver.c]: leak fixed

File: mcs/class/corlib/System/TimeZoneInfo.WebAssembly.cs [hunk 30]
```diff
@@ -5,23 +5,90 @@
...
+		extern static IntPtr mono_timezone_get_names (ref int count);
+
+		static ReadOnlyCollection<string> GetMonoWasmNames ()
+		{
+			int count = 0;
+			IntPtr array = mono_timezone_get_names (ref count);
+			if (count > 0)
+			{
+				string [] names = new string [count];
+				for (int i = 0, offset = 0; i < count; i++, offset += IntPtr.Size) {
```
Comment (kg) [inline mcs/class/corlib/System/TimeZoneInfo.WebAssembly.cs]: Is there a particular reason why we're doing this awkwardly with IntPtr and manually computing offsets, vs the alternative of just ```byte[]* pArray``` and ```pArray[i]```?

**Comment by kjpou1** [comment]
Closed this PR for preferred implementation.  https://github.com/mono/mono/pull/17760

**Pull request closed without merging**